synapse/storage/databases/main/event_federation.py

```diff
@@ -1292,6 +1292,95 @@ def _get_backfill_events(
 
         return event_id_results
 
+    @trace
+    async def record_event_failed_backfill_attempt(
+        self, room_id: str, event_id: str
+    ) -> None:
+        """
+        Record when we fail to backfill an event.
```

Comment:
> To me, "backfill an event" isn't something we do. Sometimes we try to pull individual events (via `/event`); at other times we try to backfill a section of timeline (via `/get_missing_events` or `/backfill`), starting from a particular event (but not limiting ourselves to that event). Which are we concerned with here? Or both?
> 
> Please could the terminology be clarified - in method names, docstrings, database tables, the PR description, etc - before this is merged.
> 
>  *-- @richvdh, https://github.com/matrix-org/synapse/pull/13589#pullrequestreview-1093268750*


This PR is concerned with this use case: "backfill a section of timeline (via [...] `/backfill`) starting from a particular event (but not limiting ourselves to that event)"

We pick 5 events to call `/backfill` with. Then we process the 100 events returned by `/backfill` which also probably includes the 5 we asked about. This PR is concerned with tracking when we fail to process one of these events.

Perhaps "process an event returned by backfill" is more accurate atm.

But as you're saying, there are multiple places we fetch events from a federated homeserver. It might make sense to also track when we fail fetch or process an event from `/event` or `/get_missing_events` because that means it will probably fail when we ask about it during `/backfill` as well. And maybe those other functions want to factor in the failed attempts in the future as well.

There is also the case where the event we asked about is just missing from `/backfill`. We currently don't track this case yet but I do have it as a "future PR" note in the description. It's probably not wise to ask about a event  when we never get it back.

"backfill an event" and "failed backfill attempt" still makes the most sense to me.

Reply:
can we say "backfill *from* an event then" ?

Reply:
I had that written down as well but it didn't quite seem to paint the full picture because we also track failures for events we didn't backfill from. We track failures for events we backfill from and the events returned from `/backfill` which has a "from -> to" relationship but isn't explicit.

Since we both kinda landed on it independently, I'd be willing to update it that though.

Reply:
oh, I see now. Sorry, I hadn't quite understood what it was we were tracking.

I'm afraid I still don't think it's backfilling an event. It's "processing a pulled event".

Reply:
More discussion in https://github.com/matrix-org/synapse/pull/13589#discussion_r961106420